File: java/src/min_task_time/randomTestData.java
package min_task_time;

import java.util.ArrayList;
import java.util.Random;

class randomTestData {
	private static final Random rand = new Random(0);

	int n;
	int[] time;
	int[][] before;
	int expected;

	randomTestData() {
		while (!randomTestDataSuccess());
	}

	boolean randomTestDataSuccess() {
		n = 1 + rand.nextInt(100);
		time = new int[n];
		for (var u = 0; u < n; u++) {
			time[u] = 1 + rand.nextInt(100);
		}

		var edges = new ArrayList<int[]>();
		var mat = new int[n+1][n+1];
		for (int i = 0, m = rand.nextInt(n*2); i < m; i++) {
			int u = rand.nextInt(n), v = rand.nextInt(n);
			if (mat[u][v] > 0) continue;

			edges.add(new int[] { u, v });
			mat[u][v] = time[u];
		}
		for (int u = 0; u < n; u++) {
			edges.add(new int[] { u, n });
			mat[u][n] = time[u];
		}

		for (var m = 0; m <= n; m++) {
			for (var u = 0; u <= n; u++) {
				if (mat[u][m] == 0) continue;

				for (var v = 0; v <= n; v++) {
					if (mat[m][v] == 0) continue;

					var w = mat[u][m] + mat[m][v];
					if (-w < -mat[u][v]) {
						mat[u][v] = w;
					}
				}
			}
		}

		expected = 0;
		for (var u = 0; u <= n; u++) {
			if (mat[u][u] > 0) {
				return false;
			}

			for (var v = 0; v <= n; v++) {
				if (mat[u][v] > expected) {
					expected = mat[u][v];
				}
			}
		}

		before = edges.subList(0, edges.size() - n).toArray(new int[][]{});

		return true;
	}
}

File: go/min_task_time/random_test_data.go
package min_task_time

import (
	"math/rand/v2"
)

type randomTestData struct {
	n, expected int
	time        []int
	before      [][]int
}

var random = rand.New(rand.NewPCG(0, 0))

func newRandomTestData() *randomTestData {
	var td randomTestData

	for !td.randomTestDataSuccess() {
	}

	return &td
}

func (td *randomTestData) randomTestDataSuccess() bool {
	td.n = 1 + random.IntN(100)
	td.time = make([]int, td.n)
	for u := range td.n {
		td.time[u] = 1 + random.IntN(100)
	}

	var edges [][]int
	mat := make([][]int, td.n+1)
	for i := range mat {
		mat[i] = make([]int, td.n+1)
	}

	for i, m := 0, random.IntN(td.n*2); i < m; i++ {
		u, v := random.IntN(td.n), random.IntN(td.n)
		if mat[u][v] > 0 {
			continue
		}

		edges = append(edges, []int{u, v})
		mat[u][v] = td.time[u]
	}

	for u := range td.n {
		edges = append(edges, []int{u, td.n})
		mat[u][td.n] = td.time[u]
	}

	for m := 0; m <= td.n; m++ {
		for u := 0; u <= td.n; u++ {
			if mat[u][m] == 0 {
				continue
			}

			for v := 0; v <= td.n; v++ {
				if mat[m][v] == 0 {
					continue
				}

				var w = mat[u][m] + mat[m][v]
				if -w < -mat[u][v] {
					mat[u][v] = w
				}
			}
		}
	}

	td.expected = 0
	for u := 0; u <= td.n; u++ {
		if mat[u][u] > 0 {
			return false
		}

		for v := 0; v <= td.n; v++ {
			if mat[u][v] > td.expected {
				td.expected = mat[u][v]
			}
		}
	}

	td.before = edges[:len(edges)-td.n]

	return true

}
